Hibernate Reactive thread model fixes

FroMage opened the pull request:
- Changed `Mutiny.Session Panache.getSession` to `Uni<Mutiny.Session> Panache.getSession`
- Changed `boolean PanacheEntity.isPersistent` to `Uni<Boolean> PanacheEntity.isPersistent`
- Added `@ReactiveTransaction` to forward to `Mutiny.Session.withTransaction`: method must return a `Uni` or be invoked from a non-`VertxThread` (in which case it blocks)
- Make sure we create `Mutiny.Session` objects from the right threads
- Still rely on the request-scoped `Mutiny.Session` and not the Vert.x context one used by `Mutiny.Session.withSession`
- Added `Panache.currentTransaction` which forwards to `Mutiny.Session.currentTransaction`
- Added a guide
- Fixed a mocking issue wrt interfaces

@Sanne @gavinking this should resolve the issues we've been having, right?

geoand committed:
Introduce the ability to run test methods on the vertx context

FroMage committed:
Hibernate Reactive with Panache: work on the vertx event loop

FroMage committed:
Added ReactiveTransactional

geoand committed:
Start adding asserter API

FroMage committed:
Quarkus Mockito: make sure we unwrap mocked interfaces too

FroMage committed:
Added docs

gavinking commented:
Why didn't you implement out `Context` interface?

FroMage commented:
I guess I could, but why would it be better than this?

gavinking commented:
Well ... try it?

From a superficial look at your code, I would say it probably would be simpler, but I can't be sure without trying.

FroMage commented:
Also, do we already have a configurable timeout for HR in Quarkus?

gavinking commented:
A what?

FroMage commented:
A timeout for waiting for a DB response.

Sanne commented:
> A timeout for waiting for a DB response.

no we don't have timeouts

gavinking commented:
I would not think that HR would be the right layer to do something like that.

geoand committed:
Move @RunOnVertxContext into new quarkus-junit5-vertx module

geoand committed:
Move ClassLoading shenanigans into QuarkusTestExtension

This leaves the implementations of TestMethodInvoker mostly
free of having to ClassLoading issues

geoand committed:
Introduce a proper Asserter API

geoand committed:
Activate request context on tests annotated with @RunOnVertContext

FroMage committed:
Added support for TestReactiveTransaction

FroMage committed:
Test new transactions stuff

FroMage commented:
OK, this time I've added the missing `@ReactiveTestTransaction` and rebased on main. Let's see what CI says about this.

geoand commented:
@FroMage is there anything you want me to check WRT the PR?

FroMage commented:
Yeah, changes in the test-framework from https://github.com/quarkusio/quarkus/pull/17728/commits/a69fb121ffa1f4ca8cc704ba1a435f8b73cdab54 Do you want me to extract them to their own commit?

I've added some functions, and made sure that the `finally` block that terminated the request context was moved to the end of the reactive call, and introduced something truely horrible to make sure that I could get interceptors to contribute to the asserter to contribute to the reactive pipeline.

geoand commented:
> Yeah, changes in the test-framework from [a69fb12](https://github.com/quarkusio/quarkus/commit/a69fb121ffa1f4ca8cc704ba1a435f8b73cdab54) Do you want me to extract them to their own commit?
> 
No need

> I've added some functions, 
>

👍🏼 
> made sure that the `finally` block that terminated the request context was moved to the end of the reactive call
>

very good idea, this is much better than what I had!

> introduced something truely horrible to make sure that I could get interceptors to contribute to the asserter to contribute to the reactive pipeline.

Not a thing of beauty for sure, but let's not get hung up on this details as this PR should significantly improve the user experience

geoand commented:
CI is all green 🎉

FroMage commented:
I think we should just merge this ASAP, because the API differences are breaking, so it needs to be in 2.0. Even if the way we store the session/transaction changes in the future, the API should be the same. Besides, it fixes real issues, so it's just better than what we have ATM.

geoand commented:
You have my vote

Sanne commented:
> because the API differences are breaking, so it needs to be in 2.0.

hum I see two problems with this statement :/
 - for 2.0 this is too late
 - we've being saying that we do actually not see the "2.0" major as a particular reason to break backwards compability, we can break it any time (and should always try to avoid doing so). Which is good news in the context?


> Changed boolean PanacheEntity.isPersistent to Uni<Boolean> PanacheEntity.isPersistent

Could you elaborate on why?

FroMage commented:
> for 2.0 this is too late

Ah.

> Could you elaborate on why?

Yes, since obtaining the session is now async, all operations that require it are async.

Sanne commented:
> > Could you elaborate on why?
> 
> Yes, since obtaining the session is now async, all operations that require it are async.

yea I can see that you implement this by getting a Session first, but if you don't have a session we shouldn't need to open one just for this check, right?

Which implies you don't need to change the signature.

FroMage commented:
Oh I see. I guess you're right. It can't be persistent if there isn't a session. OK lemme change that back.

Sanne commented on integration-tests/hibernate-reactive-panache/src/test/java/io/quarkus/it/panache/reactive/PanacheFunctionalityTest.java:
```diff
@@ -182,8 +199,65 @@ public void testBug7102() {
+        asserter.assertEquals(() -> Person.count(), 1l);
+    }
+
+    @DisabledOnNativeImage
```
What's wrong with native-image compatibility?

geoand replied:
In the native test case you don't have access to the running Vert.x instance where you can schedule the execution of the test method

FroMage commented:
OK done.

FroMage commented:
Ah, lemme check the docs.

FroMage commented:
Hah, the docs were wrong, so they're now correct by accident :) Brilliant!

FroMage commented:
Failure is not related: 2021-06-25T16:15:43.7464154Z Caused by: io.fabric8.maven.docker.access.hc.http.HttpRequestException: {"message":"Head https://registry-1.docker.io/v2/neo4j/neo4j-experimental/manifests/4.0.0-rc01: unknown: unable to scan account service mfa settings row from query: context deadline exceeded"} (Internal Server Error: 500)

geoand commented:
So is this good to go?

Sanne commented:
it probably is - I had asked @DavideD to play a bit more with it but I guess we can also choose to play with it after it's integrated.

geoand commented:
Up to you obviously.

Just want to mention that with `2.1` still a ways out, playing with it from `main` should not jeopardize the deliverable

FroMage commented:
I don't know what I have to do to get a review :shrug:

holomekc commented:
Hi a short question. Did you try to use hot reload while using withSession and withTransaction. We implemented an interceptor similar to ReactiveTransactionalInterceptor and in our case a hot reload destroy something, when using one of the named methods.

It fails with:
```java
java.lang.IllegalArgumentException: Type specified for TypedQuery [....entity.Agent] is incompatible with query return type [class ....entity.Agent]
	at org.hibernate.internal.AbstractSharedSessionContract.resultClassChecking(AbstractSharedSessionContract.java:878) ~[hibernate-core-5.5.3.Final.jar:5.5.3.Final]
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.createReactiveQuery(ReactiveSessionImpl.java:510) ~[hibernate-reactive-core-1.0.0.CR7.jar:1.0.0.CR7]
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.buildReactiveQueryFromName(ReactiveSessionImpl.java:487) ~[hibernate-reactive-core-1.0.0.CR7.jar:1.0.0.CR7]
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.createReactiveNamedQuery(ReactiveSessionImpl.java:472) ~[hibernate-reactive-core-1.0.0.CR7.jar:1.0.0.CR7]
	at org.hibernate.reactive.mutiny.impl.MutinySessionImpl.createNamedQuery(MutinySessionImpl.java:251) ~[hibernate-reactive-core-1.0.0.CR7.jar:1.0.0.CR7]
...
  at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54) ~[arc-2.0.0.Final.jar:2.0.0.Final]
  at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62) ~[quarkus-arc-2.0.0.Final.jar:?]
  at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49) ~[quarkus-arc-2.0.0.Final.jar:?]
  at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521) ~[quarkus:/:?]
  at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41) ~[arc-2.0.0.Final.jar:2.0.0.Final]
  at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41) ~[arc-2.0.0.Final.jar:2.0.0.Final]
  at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32) ~[arc-2.0.0.Final.jar:2.0.0.Final]
```

in general context information seems to be broken after hot reload in this case. E.g. SecurityContext crashes with NullpointerException while accessing context and some metrics which were registered at startup are also not available afterwards.

We switched to withSession and withTransaction because of this issue:
https://github.com/quarkusio/quarkus/issues/18268#issuecomment-871476405

and here the suggestion to switch to mentioned methods:
https://github.com/quarkusio/quarkus/issues/14845

We did not create an issue for this yet. But it would be great if you could check if you face similar issues in your pr.

geoand commented:
There is a very good chance the testing stuff introduced in this PR will start having conflicts soon. For that reason and because this stuff seems to be important for the user experience (not to mention there is plenty of time to iron out whatever small issues it might have), a review would be much appreciated

Sanne commented:
@geoand & @FroMage sorry I was off the last 2 days - catching up soon. Tangentially, but this happened other times as well..  and I'm sorry but I'm often dragged into urgent matters related to Hibernate, we should try to grow more people able to review Panache PRs as I can't be the one you need to wait for.

@holomekc I've noticed the same and can reproduce it without Panache as well. I've just opened #18299

geoand commented:
> @geoand & @FroMage sorry I was off the last 2 days - catching up soon. Tangentially, but this happened other times as well.. and I'm sorry but I'm often dragged into urgent matters related to Hibernate, we should try to grow more people able to review Panache PRs as I can't be the one you need to wait for.
> 

Yeah, that is totally understood :)

Sanne commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -68,27 +82,77 @@ protected abstract PanacheQueryType createPanacheQuery(Mutiny.Session session, S
+        }
+    }
+
+    private static boolean isInRequestContext(Class<?> klass) {
```
(it's of course an orthogonal issue we can address later)

FroMage replied:
:shrug:

Sanne replied:
BTW I was exploring some alternatives here but got in deep trouble with bytecode ... that's what the zulip thread is about.

Sanne reviewed (approved):
I've left some minor notes, but would prefer integrating sooner and improve later on those, if you think they are worthwhile.

Sanne commented:
Many thanks @FroMage ! merged it

edeandrea commented:
Question @FroMage - do you know if the `@ReactiveTransaction` stuff is going to make it into a `2.0.x` release (`2.0.1` hopefully)?

geoand commented:
It will almost certainly not, it's too big of a change

edeandrea commented:
Thanks @geoand

geoand commented:
With this in, I assume we can start looking into supporting Kotlin Coroutines with Hibernate Reactive Panache (cc @evanchooly)?

evanchooly commented:
That's the plan, yes.

FroMage commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
```diff
@@ -0,0 +1,1014 @@
+}
+----
+
+TIP: If you don't want to bother defining getters/setters for your entities, you can make them extend `PanacheEntityBase` and
```
We can, but by making sure everything goes via property accessors, we guarantee that they can turn the model classes from a field to a property in a backward compatible way. Not sure it's that useful in practice.

FroMage commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
```diff
@@ -0,0 +1,1014 @@
+- `set? <update-query>` will expand to `update from EntityName set <update-query>`
+
+NOTE: You can also write your queries in plain
+link:https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#hql[HQL]:
```
Yeah, probably doesn't change much.

FroMage commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
```diff
@@ -0,0 +1,1014 @@
+PanacheQuery<PersonName> query = Person.find("status", Status.Alive).project(PersonName.class);
+----
+
+1. The `@RegisterForReflection` annotation instructs Quarkus to keep the class and its members during the native compilation. More details about the `@RegisterForReflection` annotation can be found on  the xref:writing-native-applications-tips.adoc#registerForReflection[native application tips] page.
```
Well, this is only for cases where we do a projection, and since the projection class is not annotated with anything, we can't piggy-back on that annotation to detect it. We sould scan for calls to the `.project(Class)` method, now that jandex indexes use-sites, but, not sure it's worth it, this is pretty rare, non?

FroMage commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
```diff
@@ -0,0 +1,1014 @@
+
+1. The `@RegisterForReflection` annotation instructs Quarkus to keep the class and its members during the native compilation. More details about the `@RegisterForReflection` annotation can be found on  the xref:writing-native-applications-tips.adoc#registerForReflection[native application tips] page.
+2. We use public fields here, but you can use private fields and getters/setters if you prefer.
+3. This constructor will be used by Hibernate, it must be the only constructor in your class and have all the class attributes as parameters.
```
Not sure. https://docs.jboss.org/hibernate/orm/5.5/userguide/html_single/Hibernate_User_Guide.html#hql-select-clause says "The projection class must be fully qualified in the entity query, and it must define a matching constructor." Perhaps we can clarify this.

FroMage commented on docs/src/main/asciidoc/hibernate-reactive-panache.adoc:
```diff
@@ -0,0 +1,1014 @@
+
+Hibernate Reactive with Panache relies on compile-time bytecode enhancements to your entities.
+
+It attempts to identity archives with Panache entities (and consumers of Panache entities)
```
Yup

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -68,27 +82,77 @@ protected abstract PanacheQueryType createPanacheQuery(Mutiny.Session session, S
+        if (io.vertx.core.Context.isOnVertxThread()) {
+            return Uni.createFrom().item(Arc.container().instance(Mutiny.Session.class).get());
+        } else {
+            // FIXME: we may need context propagation
```
Well context propagation, using MP-CP, but I'd rather delay that until someone reports it.
